common/components/org.wso2.carbon.data.agents.log4j/src/main/java/org/wso2/carbon/data/agents/log4j/appender/LogEventAppender.java

```diff
@@ -0,0 +1,411 @@
+                            getMaxTolerableConsecutiveFailure() + ". Purging log event array. Some logs will be lost.");
+                    loggingEvents.clear();
+                    numOfConsecutiveFailures = 0;
+                }
+            }
+        }
+
+        private void publishLogEvent(TenantAwareLoggingEvent event)
+                throws ParseException, MalformedURLException, AuthenticationException, TransportException,
+                MalformedStreamDefinitionException, StreamDefinitionException,
+                DifferentStreamDefinitionAlreadyDefinedException, ExecutionException {
+            String streamId = "";
+            String streamName = "";
+            String tenantId = event.getTenantId();
+            if (tenantId.equals(String.valueOf(MultitenantConstants.INVALID_TENANT_ID)) || tenantId
+                    .equals(String.valueOf(MultitenantConstants.SUPER_TENANT_ID))) {
+                tenantId = "0";
+            }
+            String serverKey = getCurrentServerName();
+            String currDateStr = getCurrentDate();
+            try {
+                Pattern pattern = Pattern.compile("(tcp):\\/\\/([a-zA-Z0-9]+):([0-9]+)");
+                Matcher matcher = pattern.matcher(url);
+                if (matcher.find() && authURLSet == null) {
+                    authURLSet = "ssl://" + matcher.group(2).toString() + ":" + (
+                            Integer.parseInt(matcher.group(3).toString()) + 100);
+                }
+                dataPublisher = new DataPublisher("Thrift", url, authURLSet, userName, password);
+            } catch (DataEndpointAgentConfigurationException e) {
+                log.error(
+                        "Error in initializing binary data-publisher to send requests to global throttling engine " + e
+                                .getMessage(), e);
+            } catch (DataEndpointException e) {
+                log.error(
+                        "Error in initializing binary data-publisher to send requests to global throttling engine " + e
+                                .getMessage(), e);
+            } catch (DataEndpointConfigurationException e) {
+                log.error(
+                        "Error in initializing binary data-publisher to send requests to global throttling engine " + e
+                                .getMessage(), e);
+            } catch (DataEndpointAuthenticationException e) {
+                log.error(
+                        "Error in initializing binary data-publisher to send requests to global throttling engine " + e
+                                .getMessage(), e);
+            } catch (TransportException e) {
+                log.error(
+                        "Error in initializing binary data-publisher to send requests to global throttling engine " + e
+                                .getMessage(), e);
+            }
```

Comment:
I think even with this approach, you can't prevent the NPE log appearing all the time. New log events will still be put onto the log-event queue regardless of purging when no-of-retry exceeds. Adding to queue and removing from queue are happening with two different threads and there will always be new log evens added to the queue and you will still observe NPEs. I wonder why this is not observed with your test cases. 

Can you explain why we will not observe NPE logs multiple times, if the server is kept on running and new logs events are put on to the queue?